script.js

```diff
@@ -38,30 +47,122 @@ function shuffle(array) {
 
 let shuffledColors = shuffle(COLORS);
 
-// this function loops over the array of colors
-// it creates a new div and gives it a class with the value of the color
-// it also adds an event listener for a click for each card
 function createDivsForColors(colorArray) {
   for (let color of colorArray) {
-    // create a new div
     const newDiv = document.createElement("div");
 
-    // give it a class attribute for the value we are looping over
     newDiv.classList.add(color);
-
-    // call a function handleCardClick when a div is clicked on
     newDiv.addEventListener("click", handleCardClick);
 
-    // append the div to the element with an id of game
     gameContainer.append(newDiv);
   }
 }
 
-// TODO: Implement this function!
+function resetOldGame() {
+  setScore(0);
+  document.querySelector("#current-score").innerText = "";
+  document.querySelector("#final-score").innerText = "";
+  document.querySelector("#game-over").innerText = "";
+  document.querySelector("#play-again").innerText = "";
+}
+
+function initializeNewGame() {
+  let game = document.querySelector("#game");
+  game.innerHTML = "";
+  let newDeck = shuffle(COLORS);
+  createDivsForColors(newDeck);
+}
+
+function startGame() {
+  resetOldGame();
+  initializeNewGame();
+}
+
+const chooseCards = function(chosenCard) {
+  if (firstCard === null) {
+    firstCard = chosenCard;
+    setScore(currentScore + 1);
+  } else if (secondCard === null) {
+    secondCard = chosenCard;
+    setScore(currentScore + 1);
+  }
+
+  if (firstCard && secondCard) {
+    preventClicks = true;
+  }
+  
+  chosenCard.classList.add("revealed");
+  chosenCard.style.backgroundColor = chosenCard.classList[0];
+}
+
+const checkForMatch = function() {
+  if (firstCard.className === secondCard.className) {
+    revealedCards += 2;
+    removeEventListener();
+    preventClicks = false;
+  } else {
+    setTimeout(function() {
+      forgetChosenCards();
+      preventClicks = false;
+    }, 1000);
+  }
+}
+
+const removeEventListener = function() {
+  firstCard.removeEventListener('click', handleCardClick);
+  secondCard.removeEventListener('click', handleCardClick);
+  firstCard = null;
+  secondCard = null;
+}
+
+const forgetChosenCards = function() {
+  firstCard.style.backgroundColor = null;
+  firstCard.classList.remove("revealed");
+  firstCard = null;
+
+  secondCard.style.backgroundColor = null;
+  secondCard.classList.remove("revealed");
+  secondCard = null;
+}
+
 function handleCardClick(event) {
```

Comment:
clean function implementation, nice!

Reply:
Thank you. As I mentioned, I do have a few places where I could DRY up the functions, but when I tried, it caused issues. I believe that's because they were not stateless; I have some ideas for how I could do that, but thought it was not essential at this first pass. I'll explain more when we meet.